cmd/thanos/rule.go

```diff
@@ -767,3 +763,63 @@ func removeDuplicateQueryAddrs(logger log.Logger, duplicatedQueriers prometheus.
 	}
 	return deduplicated
 }
+
+// queryFunc returns query function that hits the HTTP query API of query peers in randomized order until we get a result
+// back or the context get canceled.
+func queryFunc(
+	logger log.Logger,
+	peer cluster.Peer,
+	dnsProvider *dns.Provider,
+	duplicatedQuery prometheus.Counter,
+	partialResponseDisabled bool,
+) rules.QueryFunc {
+	const partialResponseParam = "partial_response"
+
+	partialResponseValue := strconv.FormatBool(!partialResponseDisabled)
+	spanID := "/rule_instant_query HTTP[client]"
+	if partialResponseDisabled {
+		spanID = "/rule_instant_query_strict HTTP[client]"
+	}
+
+	return func(ctx context.Context, q string, t time.Time) (promql.Vector, error) {
+		var addrs []string
+
+		// Add addresses from gossip.
+		peers := peer.PeerStates(cluster.PeerTypeQuery)
+		var ids []string
+		for id := range peers {
+			ids = append(ids, id)
+		}
+		sort.Slice(ids, func(i int, j int) bool {
+			return strings.Compare(ids[i], ids[j]) < 0
+		})
+		for _, id := range ids {
+			addrs = append(addrs, peers[id].QueryAPIAddr)
+		}
+
+		// Add DNS resolved addresses from static flags and file SD.
+		// TODO(bwplotka): Consider generating addresses in *url.URL
+		addrs = append(addrs, dnsProvider.Addresses()...)
+
+		removeDuplicateQueryAddrs(logger, duplicatedQuery, addrs)
+
+		for _, i := range rand.Perm(len(addrs)) {
+			u, err := url.Parse(fmt.Sprintf("http://%s", addrs[i]))
+			if err != nil {
+				return nil, errors.Wrapf(err, "url parse %s", addrs[i])
+			}
+			form, err := url.ParseQuery(u.RawQuery)
+			if err != nil {
+				return nil, errors.Wrapf(err, "url raw query parse %s", addrs[i])
+			}
+			form.Add(partialResponseParam, partialResponseValue)
+			u.RawQuery = form.Encode()
+
+			span, ctx := tracing.StartSpan(ctx, spanID)
+			v, err := promclient.PromqlQueryInstant(ctx, logger, u, q, t, true)
```

Comment:
It would be useful to at least log or maybe even increment counter about partial response to the rule/alert evaluation?

Reply:
Also It would be great to see this in the `rules` page that last evaluation was partial result possibly

Reply:
will add TODO